algorithms/src/snark/marlin/ahp/ahp.rs

```diff
@@ -258,67 +352,165 @@ impl<F: PrimeField, MM: MarlinMode> AHPForR1CS<F, MM> {
 
         Ok(linear_combinations)
     }
 
     #[allow(clippy::too_many_arguments)]
-    fn construct_lhs(
-        label: &str,
+    fn add_g_m_term<E: EvaluationsProvider<F>>(
+        matrix_sumcheck: &mut LinearCombination<F>,
+        id: CircuitId,
+        m: &str,
         alpha: F,
         beta: F,
         gamma: F,
-        v_h_at_alpha_beta: F,
-        g_at_gamma: F,
+        evals: &E,
+        g_m: &LinearCombination<F>,
+        v_h_i_alpha_beta: F,
+        r: F,
         sum: F,
+        selector: F,
+    ) -> Result<(), AHPError> {
+        let g_at_gamma = evals.get_lc_eval(g_m, gamma)?;
+        let b_term = gamma * g_at_gamma + sum; // Xg_m(X) + sum_m
+        let lhs_a = Self::construct_lhs(id, m, alpha, beta, v_h_i_alpha_beta, b_term, selector);
+        *matrix_sumcheck += (r, &lhs_a);
+        Ok(())
+    }
+
+    #[allow(clippy::too_many_arguments)]
+    fn construct_lhs(
+        id: CircuitId,
+        label: &str,
+        alpha: F,
+        beta: F,
+        v_h_i_at_alpha_beta: F,
+        b_term: F,
         selector_at_gamma: F,
     ) -> LinearCombination<F> {
-        let a =
-            LinearCombination::new("a_poly_".to_string() + label, [(v_h_at_alpha_beta, "val_".to_string() + label)]);
+        let label_row = format!("circuit_{id}_row_{label}");
+        let label_col = format!("circuit_{id}_col_{label}");
+        let label_val = format!("circuit_{id}_val_{label}");
+        let label_row_col = format!("circuit_{id}_row_col_{label}");
+        let label_a_poly = format!("circuit_{id}_a_poly_{label}");
+        let label_denom = format!("circuit_{id}_denom_{label}");
+
+        let a = LinearCombination::new(label_a_poly, [(v_h_i_at_alpha_beta, label_val)]);
         let alpha_beta = alpha * beta;
 
-        let mut b = LinearCombination::new("denom_".to_string() + label, [
+        let mut b = LinearCombination::new(label_denom, [
             (alpha_beta, LCTerm::One),
-            (-alpha, ("row_".to_string() + label).into()),
-            (-beta, ("col_".to_string() + label).into()),
-            (F::one(), ("row_col_".to_string() + label).into()),
+            (-alpha, (label_row).into()),
+            (-beta, (label_col).into()),
+            (F::one(), (label_row_col).into()),
         ]);
-        b *= gamma * g_at_gamma + sum;
+        b *= b_term; // Xg_m(X) + sum_m
 
         let mut lhs = a;
         lhs -= &b;
         lhs *= selector_at_gamma;
         lhs
     }
+
+    fn get_selector_evaluation(
+        cached_selector_evaluations: &mut BTreeMap<(u64, u64, F), F>,
+        largest_domain: &EvaluationDomain<F>,
+        target_domain: &EvaluationDomain<F>,
+        challenge: F,
+    ) -> F {
+        *cached_selector_evaluations
+            .entry((target_domain.size, largest_domain.size, challenge))
+            .or_insert_with(|| largest_domain.evaluate_selector_polynomial(*target_domain, challenge))
+    }
```

Comment:
I would just make this a free function named `selector_evals`, so that calling it becomes significantly shorter: `selector_evals(...)` instead of `Self::get_selector_evaluations(...)`.

Even better would be if we we could add it as a method to `BTreeMap`, but that seems like work that's not worth it.

Reply:
Nice, fixed in b17840352